### PR TITLE
[Loader] fix nested-component loading

### PR DESCRIPTION
**Pull request by dannypsnl**

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

File: lib/loader/ast/section.cpp [hunk 1]
```diff
@@ -217,26 +218,38 @@ Expect<void> Loader::loadSection(AST::TagSection &Sec) {
...
+    }
+    auto NestedComp = std::make_shared<AST::Component::Component>();
+    NestedComp->getMagic() = WasmMagic;
+    NestedComp->getVersion() = {Ver[0], Ver[1]};
+    NestedComp->getLayer() = {Ver[2], Ver[3]};
+
+    auto Offset = FMgr.getOffset();
+
+    if (auto Res =
+            loadComponent(*NestedComp, ExpectedSize - (Offset - StartOffset));
```
Comment (q82419) [inline lib/loader/ast/section.cpp:244]: I think there may be a situation that `ExpectedSize` < `Offset` - `StartOffset`?
Consider a situation that the section size was loaded and set as size `10` in `Loader::loadSectionContent<T>()`, and `StartOffset` is `50` when entering this function, and the `Offset` is `70` after above loading. Then the evaluate of `ExpectedSize - (Offset - StartOffset)` will underflow in a `uint32_t`.

**Review by q82419 — CHANGES_REQUESTED** [review]

File: lib/loader/ast/section.cpp (hunk 1 above)
Comment (dannypsnl) [reply lib/loader/ast/section.cpp:249]: Indeed, I push a fix for this, there are two place use the same idea. (nested module and component)

File: lib/loader/ast/section.cpp [hunk 2]
```diff
@@ -217,26 +218,43 @@ Expect<void> Loader::loadSection(AST::TagSection &Sec) {
...
+                          FMgr.getLastOffset(), ASTNodeAttr::Component);
+    }
+    auto NestedComp = std::make_shared<AST::Component::Component>();
+    NestedComp->getMagic() = WasmMagic;
+    NestedComp->getVersion() = {Ver[0], Ver[1]};
+    NestedComp->getLayer() = {Ver[2], Ver[3]};
+
+    auto Offset = FMgr.getOffset();
+
+    if (ExpectedSize < Offset - StartOffset) {
```
Comment (q82419) [inline lib/loader/ast/section.cpp:243]: I think this is suit for `unlikely`.

File: lib/loader/ast/section.cpp [hunk 3]
```diff
@@ -259,9 +277,15 @@ Expect<void> Loader::loadSection(AST::CoreModuleSection &Sec) {
     CoreMod.getVersion() = Ver;
 
     auto Offset = FMgr.getOffset();
-    ExpectedSize -= (Offset - StartOffset);
 
-    if (auto Res = loadModuleInBound(CoreMod, ExpectedSize); !Res) {
+    if (ExpectedSize < Offset - StartOffset) {
```
Comment (q82419) [inline lib/loader/ast/section.cpp:281]: I think this is suit for `unlikely`.

**Review by q82419 — CHANGES_REQUESTED** [review]

**Review by q82419 — APPROVED** [review]

**Pull request merged**